### PR TITLE
Add MulAccOps trait for Montgomery multiplication support

### PR DESCRIPTION
**Pull request by kaidokert**
Provides algorithm-agnostic row-level fused multiply-accumulate primitives                                                                                                      
(mul_acc_row, mul_acc_shift_row) that abstract over the internal limb                                                                                                           
layout. Higher-level crates (e.g. modmath) use these to implement CIOS or                                                                                                       
other Montgomery multiplication variants without touching raw word arrays.  

## Summary by Sourcery

Introduce a CIOS Montgomery multiplication support trait and implement it for FixedUInt to expose row-level fused multiply-accumulate operations without leaking internal limb representation.

New Features:
- Add a CiosOps trait defining row-level fused multiply-accumulate and conditional subtraction operations for CIOS Montgomery multiplication.
- Implement the CiosOps trait for FixedUInt, providing word access, multiply-accumulate row operations, and conditional modular subtraction.
- Re-export the CiosOps trait from the crate root for use by higher-level crates.

Tests:
- Add unit tests covering CiosOps word access, default behavior, row multiply-accumulate operations (with and without overflow), shift-row operations, and conditional subtraction semantics for FixedUInt.

## Summary by Sourcery

Introduce a generic MulAccOps trait for row-level fused multiply-accumulate operations and implement it for FixedUInt to support Montgomery-style multiplication without exposing internal limb layouts.

New Features:
- Add a MulAccOps trait defining word-level fused multiply-accumulate and shift-row operations for big-integer types used in Montgomery multiplication algorithms.
- Implement MulAccOps for FixedUInt, including word-count and word-access helpers, enabling scalar-by-row accumulation primitives over its internal limbs.
- Re-export MulAccOps from the crate root so higher-level crates can depend on these operations without coupling to FixedUInt internals.

Tests:
- Add unit tests validating MulAccOps word access, zero-initialization, basic and overflowing mul-accumulate rows, shift-row behavior (including carry and overflow handling), and edge cases over different FixedUInt sizes.

## Summary by CodeRabbit

* **New Features**
  * Added a public MulAccOps API exposing word-level access, row multiply‑accumulate, shifted-row operations, and overflow reporting.
  * Exposed the new module and re‑exported MulAccOps so it’s available to users.
  * Provided an implementation for the existing fixed-size big‑integer type enabling these operations.

* **Tests**
  * Added comprehensive tests covering word access, accumulation, shifted behavior, overflow cases, and edge conditions.

### CODE REVIEW

File: src/fixeduint/cios_ops_impl.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+where
+    T: MachineWord
+        + CarryingMul<Output = T>
+        + ConstCarryingAdd
+        + ConstBorrowingSub
+        + num_traits::Zero
+        + num_traits::One
+        + num_traits::WrappingMul
+        + num_traits::ops::overflowing::OverflowingAdd
+        + core::ops::Add<Output = T>,
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint/cios_ops_impl.rs]: done

File: src/fixeduint/cios_ops_impl.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+{
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint/cios_ops_impl.rs]: Mostly done ?

File: src/fixeduint/mul_acc_ops_impl.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+    fn mul_acc_shift_row(scalar: T, multiplicand: &Self, acc: &mut Self, acc_hi: T) -> T {
+        debug_assert!(N > 0, "MulAccOps requires at least one word");
+        // First word: discarded (zero by construction)
+        let (_, mut carry) = CarryingMul::carrying_mul_add(
+            scalar,
+            multiplicand.array[0],
+            acc.array[0],
+            <T as ConstZero>::zero(),
+        );
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply src/fixeduint/mul_acc_ops_impl.rs]: Yeah i've been nagged about this N=0 a thousand times here. Lets just accept that there's a) no Rust stable compatible way to do this at the type level, there's no compile-time enforcement possible b) these N=0 checks could be in constructors and From or from_ constructors at runtime, but it would clutter everything a lot and it's hard to enforce uniformly.

Maybe lets just leave it for a future const / compiler fix that allows compile-time assertions on these things ?

File: src/mul_acc_ops.rs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+///
+/// Implementors provide scalar × bigint accumulate/shift primitives.
+/// A higher-level crate (e.g. modmath) orchestrates the outer loop of
+/// a Montgomery multiplication variant using only these methods, without
+/// knowledge of the internal representation.
+pub trait MulAccOps: Sized + Copy + Default + crate::const_numtraits::ConstZero {
+    /// The machine-word type used as a scalar in row operations.
+    type Word: Copy
+        + crate::const_numtraits::ConstZero
+        + crate::const_numtraits::ConstOne
```
Comment (SourceryAI) [inline src/mul_acc_ops.rs:18]: **suggestion:** Consider whether all current trait bounds on `Word` are necessary for this API surface.

`Word` is currently bounded by `ConstZero`, `ConstOne`, and `PartialOrd`, but these constraints are not used in `MulAccOps` itself. If callers don’t actually depend on these traits, you could drop them to support more `Word` types and avoid over-constraining implementations. If they are required by specific consumers, consider capturing that via a separate alias trait so the dependency is explicit and localized.

Suggested implementation:

```rust
    /// The machine-word type used as a scalar in row operations.
    ///
    /// Note: this is intentionally minimally constrained to avoid
    /// over-constraining implementors. Call sites that require
    /// `ConstZero`, `ConstOne`, `PartialOrd`, etc. should add those
    /// bounds on their own generics or via an alias trait.
    type Word: Copy;

```

To fully implement the suggestion you’ll want to:
1. Audit all uses of `MulAccOps::Word` (and any generic parameters constrained by `MulAccOps`) and:
   - Add explicit bounds (`ConstZero`, `ConstOne`, `PartialOrd`, etc.) where they are actually required, e.g.:
     - `fn foo<O: MulAccOps>(...) where O::Word: ConstZero + ConstOne + PartialOrd { ... }`
   - Or introduce an alias trait, e.g. `pub trait OrderedWord: ConstZero + ConstOne + PartialOrd {}` and use it in those consumers instead of baking the bounds into `MulAccOps`.
2. If any inherent impls or blanket impls rely on the stronger bounds (e.g., `impl MulAccOps for T where T::Word: ConstZero + ...`), move those bounds into `where`-clauses on the impls instead of the `MulAccOps` definition.
3. Run the test suite and fix any compilation errors by tightening bounds only at the call sites that truly depend on those traits.

**Review by SourceryAI — COMMENTED** [review]
Hey - I've found 1 issue, and left some high level feedback:

- In the `MulAccOps` trait, the `Word` associated type is constrained with `ConstOne` and `PartialOrd` but these bounds are not used by the trait methods; consider removing them to keep the API surface minimal and make it easier to implement the trait for other bigint types.
- The `mul_acc_shift_row` implementation assumes `N > 0` and only guards this with a `debug_assert!`; if a zero-limb `FixedUInt<_, 0>` were ever constructed this would be UB in release builds, so it may be safer to prevent `MulAccOps` from being implemented for `N = 0` at the type level or to add a runtime check/panic for that case.

<details>
<summary>Prompt for AI Agents</summary>

~~~markdown
Please address the comments from this code review:

## Overall Comments
- In the `MulAccOps` trait, the `Word` associated type is constrained with `ConstOne` and `PartialOrd` but these bounds are not used by the trait methods; consider removing them to keep the API surface minimal and make it easier to implement the trait for other bigint types.
- The `mul_acc_shift_row` implementation assumes `N > 0` and only guards this with a `debug_assert!`; if a zero-limb `FixedUInt<_, 0>` were ever constructed this would be UB in release builds, so it may be safer to prevent `MulAccOps` from being implemented for `N = 0` at the type level or to add a runtime check/panic for that case.

## Individual Comments

### Comment 1
<location path="src/mul_acc_ops.rs" line_range="16-18" />
<code_context>
+/// knowledge of the internal representation.
+pub trait MulAccOps: Sized + Copy + Default + crate::const_numtraits::ConstZero {
+    /// The machine-word type used as a scalar in row operations.
+    type Word: Copy
+        + crate::const_numtraits::ConstZero
+        + crate::const_numtraits::ConstOne
+        + PartialOrd;
+
</code_context>
<issue_to_address>
**suggestion:** Consider whether all current trait bounds on `Word` are necessary for this API surface.

`Word` is currently bounded by `ConstZero`, `ConstOne`, and `PartialOrd`, but these constraints are not used in `MulAccOps` itself. If callers don’t actually depend on these traits, you could drop them to support more `Word` types and avoid over-constraining implementations. If they are required by specific consumers, consider capturing that via a separate alias trait so the dependency is explicit and localized.

Suggested implementation:

```rust
    /// The machine-word type used as a scalar in row operations.
    ///
    /// Note: this is intentionally minimally constrained to avoid
    /// over-constraining implementors. Call sites that require
    /// `ConstZero`, `ConstOne`, `PartialOrd`, etc. should add those
    /// bounds on their own generics or via an alias trait.
    type Word: Copy;

```

To fully implement the suggestion you’ll want to:
1. Audit all uses of `MulAccOps::Word` (and any generic parameters constrained by `MulAccOps`) and:
   - Add explicit bounds (`ConstZero`, `ConstOne`, `PartialOrd`, etc.) where they are actually required, e.g.:
     - `fn foo<O: MulAccOps>(...) where O::Word: ConstZero + ConstOne + PartialOrd { ... }`
   - Or introduce an alias trait, e.g. `pub trait OrderedWord: ConstZero + ConstOne + PartialOrd {}` and use it in those consumers instead of baking the bounds into `MulAccOps`.
2. If any inherent impls or blanket impls rely on the stronger bounds (e.g., `impl MulAccOps for T where T::Word: ConstZero + ...`), move those bounds into `where`-clauses on the impls instead of the `MulAccOps` definition.
3. Run the test suite and fix any compilation errors by tightening bounds only at the call sites that truly depend on those traits.
</issue_to_address>
~~~

</details>

***

Hi @kaidokert! 👋

Thanks for trying out Sourcery by commenting with `@sourcery-ai review`! 🚀

Install the [sourcery-ai bot](https://github.com/apps/sourcery-ai/installations/new) to get automatic code reviews on every pull request ✨

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment and I'll use the feedback to improve your reviews.
</sub>
